File: backend/app/services/rbac_constraint_service.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.rbac_constraint import RBACConstraint, RBACConstraintRoleAssociation
from app.models.rbac_role import RBACRole
from app.models.rbac_user_role_assignment import RBACUserRoleAssignment
from app.services.rbac_service import get_role_descendants


async def validate_role_hierarchy(
    parent_role_id: str, child_role_id: str, db: AsyncSession
) -> bool:
    """Return True if adding parent -> child would not create a cycle.

    A role cannot be its own parent, and a child cannot already be a
    descendant of the proposed parent.
    """
    if parent_role_id == child_role_id:
        return False

    descendants = await get_role_descendants(parent_role_id, db)
    descendant_ids = {role.id for role in descendants}
    return child_role_id not in descendant_ids


async def validate_user_role_assignments(
    user_id: str, proposed_role_ids: set[str], db: AsyncSession
) -> list[str]:
    """Return human-readable constraint violations for a proposed role set.

    The effective role set merges the user's currently valid assignments with
    ``proposed_role_ids``. Callers that are replacing roles should pass the
    full replacement set; callers that are adding roles should pass the union
    of existing and new role IDs.
    """
    now = datetime.utcnow()

    active_constraints_result = await db.execute(
        select(RBACConstraint).where(RBACConstraint.is_active.is_(True))
    )
    active_constraints = active_constraints_result.scalars().all()
    if not active_constraints:
        return []

    constraint_ids = {constraint.id for constraint in active_constraints}
    associations_result = await db.execute(
        select(RBACConstraintRoleAssociation).where(
            RBACConstraintRoleAssociation.constraint_id.in_(constraint_ids)
        )
    )
    associations = associations_result.scalars().all()

    associations_by_constraint: dict[str, dict[str, set[str]]] = {}
    for association in associations:
        by_type = associations_by_constraint.setdefault(association.constraint_id, {})
        by_type.setdefault(association.association_type, set()).add(association.role_id)

    existing_assignments_result = await db.execute(
        select(RBACUserRoleAssignment).where(
            RBACUserRoleAssignment.user_id == user_id,
            (
                RBACUserRoleAssignment.valid_from.is_(None)
                | (RBACUserRoleAssignment.valid_from <= now)
            ),
            (
                RBACUserRoleAssignment.valid_until.is_(None)
                | (RBACUserRoleAssignment.valid_until >= now)
            ),
        )
    )
    existing_role_ids = {
        assignment.role_id for assignment in existing_assignments_result.scalars().all()
    }
    effective_role_ids = existing_role_ids | proposed_role_ids

    all_relevant_role_ids: set[str] = set(effective_role_ids)
    for constraint in active_constraints:
        for role_id_set in associations_by_constraint.get(constraint.id, {}).values():
            all_relevant_role_ids.update(role_id_set)

    role_name_map: dict[str, str] = {}
    if all_relevant_role_ids:
        role_rows_result = await db.execute(
            select(RBACRole.id, RBACRole.name).where(
                RBACRole.id.in_(all_relevant_role_ids)
            )
        )
        role_name_map = {
            row[0]: row[1] for row in role_rows_result.all() if row[1] is not None
        }

    violations: list[str] = []

    for constraint in active_constraints:
        groups = associations_by_constraint.get(constraint.id, {})
        subject_role_ids = groups.get("subject", set())
        prerequisite_role_ids = groups.get("prerequisite", set())

        if constraint.constraint_type == "mutual_exclusive":
            conflict_ids = subject_role_ids & effective_role_ids
            if len(conflict_ids) >= 2:
                names = _join_role_names(sorted(conflict_ids), role_name_map)
                violations.append(f"角色 {names} 互斥，不能同时分配")

        elif constraint.constraint_type == "prerequisite":
            missing_prerequisite_ids = prerequisite_role_ids - effective_role_ids
            if missing_prerequisite_ids:
                for subject_id in subject_role_ids & effective_role_ids:
                    subject_name = role_name_map.get(subject_id, subject_id)
                    missing_names = _join_role_names(
                        sorted(missing_prerequisite_ids), role_name_map
                    )
                    violations.append(
                        f"拥有角色 {subject_name} 需要先拥有角色 {missing_names}"
                    )

        elif constraint.constraint_type == "cardinality":
            config = constraint.config or {}
            max_users = config.get("max_users")
            if max_users is None or not subject_role_ids:
                continue

            current_users_result = await db.execute(
                select(RBACUserRoleAssignment.user_id)
                .where(
                    RBACUserRoleAssignment.role_id.in_(subject_role_ids),
                    (
                        RBACUserRoleAssignment.valid_from.is_(None)
                        | (RBACUserRoleAssignment.valid_from <= now)
                    ),
                    (
                        RBACUserRoleAssignment.valid_until.is_(None)
                        | (RBACUserRoleAssignment.valid_until >= now)
                    ),
                )
                .distinct()
            )
            current_user_ids = set(current_users_result.scalars().all())

            user_already_counted = user_id in current_user_ids
            would_receive_subject = bool(subject_role_ids & proposed_role_ids)
            projected_count = len(current_user_ids) + (
                1 if would_receive_subject and not user_already_counted else 0
            )

            if projected_count > max_users:
                names = _join_role_names(sorted(subject_role_ids), role_name_map)
                violations.append(f"角色 {names} 最多只能分配给 {max_users} 个用户")

    return violations


def _join_role_names(role_ids: list[str], role_name_map: dict[str, str]) -> str:
    names = [role_name_map.get(role_id, role_id) for role_id in role_ids]
    return "、".join(names)

File: backend-go/services/rbac_constraint_service.go
package services

import (
	"encoding/json"
	"sort"
	"strconv"
	"strings"
	"time"

	"ai-multi-platform-release/backend-go/models"
)

// ValidateRoleHierarchy returns true if adding parentRoleID -> childRoleID
// would not create a cycle.
func ValidateRoleHierarchy(parentRoleID, childRoleID string) (bool, error) {
	if parentRoleID == childRoleID {
		return false, nil
	}
	descendants, err := GetRoleDescendants(childRoleID)
	if err != nil {
		return false, err
	}
	for _, d := range descendants {
		if d.ID == parentRoleID {
			return false, nil
		}
	}
	return true, nil
}

// ValidateUserRoleAssignments returns human-readable constraint violations for
// a proposed role set. The effective role set merges the user's currently
// valid assignments with proposedRoleIDs.
func ValidateUserRoleAssignments(userID string, proposedRoleIDs map[string]bool) ([]string, error) {
	o := GetOrm()
	now := time.Now()

	var activeConstraints []models.RBACConstraint
	_, err := o.QueryTable(new(models.RBACConstraint)).
		Filter("is_active", true).
		All(&activeConstraints)
	if err != nil {
		return nil, err
	}
	if len(activeConstraints) == 0 {
		return nil, nil
	}

	constraintIDs := make([]string, 0, len(activeConstraints))
	for _, c := range activeConstraints {
		constraintIDs = append(constraintIDs, c.ID)
	}

	var associations []models.RBACConstraintRoleAssociation
	_, err = o.QueryTable(new(models.RBACConstraintRoleAssociation)).
		Filter("constraint_id__in", constraintIDs).
		All(&associations)
	if err != nil {
		return nil, err
	}

	associationsByConstraint := map[string]map[string]map[string]bool{}
	for _, a := range associations {
		byType, ok := associationsByConstraint[a.ConstraintID]
		if !ok {
			byType = map[string]map[string]bool{}
			associationsByConstraint[a.ConstraintID] = byType
		}
		roleSet, ok := byType[a.AssociationType]
		if !ok {
			roleSet = map[string]bool{}
			byType[a.AssociationType] = roleSet
		}
		roleSet[a.RoleID] = true
	}

	var existingAssignments []models.RBACUserRoleAssignment
	_, err = o.QueryTable(new(models.RBACUserRoleAssignment)).
		Filter("user_id", userID).
		All(&existingAssignments)
	if err != nil {
		return nil, err
	}
	existingRoleIDs := map[string]bool{}
	for _, a := range existingAssignments {
		if a.ValidFrom != nil && a.ValidFrom.After(now) {
			continue
		}
		if a.ValidUntil != nil && a.ValidUntil.Before(now) {
			continue
		}
		existingRoleIDs[a.RoleID] = true
	}

	effectiveRoleIDs := map[string]bool{}
	for id := range existingRoleIDs {
		effectiveRoleIDs[id] = true
	}
	for id := range proposedRoleIDs {
		effectiveRoleIDs[id] = true
	}

	allRelevantRoleIDs := map[string]bool{}
	for id := range effectiveRoleIDs {
		allRelevantRoleIDs[id] = true
	}
	for _, constraint := range activeConstraints {
		for _, roleSet := range associationsByConstraint[constraint.ID] {
			for id := range roleSet {
				allRelevantRoleIDs[id] = true
			}
		}
	}

	roleNameMap := map[string]string{}
	if len(allRelevantRoleIDs) > 0 {
		var roles []models.RBACRole
		ids := make([]string, 0, len(allRelevantRoleIDs))
		for id := range allRelevantRoleIDs {
			ids = append(ids, id)
		}
		_, err = o.QueryTable(new(models.RBACRole)).
			Filter("id__in", ids).
			All(&roles)
		if err != nil {
			return nil, err
		}
		for _, r := range roles {
			if r.Name != "" {
				roleNameMap[r.ID] = r.Name
			}
		}
	}

	var violations []string

	for _, constraint := range activeConstraints {
		groups := associationsByConstraint[constraint.ID]
		subjectRoleIDs := groups["subject"]
		prerequisiteRoleIDs := groups["prerequisite"]

		switch constraint.ConstraintType {
		case "mutual_exclusive":
			conflictIDs := []string{}
			for id := range subjectRoleIDs {
				if effectiveRoleIDs[id] {
					conflictIDs = append(conflictIDs, id)
				}
			}
			if len(conflictIDs) >= 2 {
				sort.Strings(conflictIDs)
				violations = append(violations, "角色 "+joinRoleNames(conflictIDs, roleNameMap)+" 互斥，不能同时分配")
			}

		case "prerequisite":
			missingPrerequisiteIDs := []string{}
			for id := range prerequisiteRoleIDs {
				if !effectiveRoleIDs[id] {
					missingPrerequisiteIDs = append(missingPrerequisiteIDs, id)
				}
			}
			if len(missingPrerequisiteIDs) > 0 {
				sortedSubject := make([]string, 0, len(subjectRoleIDs))
				for id := range subjectRoleIDs {
					if effectiveRoleIDs[id] {
						sortedSubject = append(sortedSubject, id)
					}
				}
				sort.Strings(sortedSubject)
				sort.Strings(missingPrerequisiteIDs)
				for _, subjectID := range sortedSubject {
					subjectName := roleNameMap[subjectID]
					if subjectName == "" {
						subjectName = subjectID
					}
					missingNames := joinRoleNames(missingPrerequisiteIDs, roleNameMap)
					violations = append(violations, "拥有角色 "+subjectName+" 需要先拥有角色 "+missingNames)
				}
			}

		case "cardinality":
			config := map[string]interface{}{}
			if constraint.Config != "" {
				_ = json.Unmarshal([]byte(constraint.Config), &config)
			}
			maxUsers, ok := config["max_users"].(float64)
			if !ok || len(subjectRoleIDs) == 0 {
				continue
			}

			var assigned []models.RBACUserRoleAssignment
			subjectIDs := make([]string, 0, len(subjectRoleIDs))
			for id := range subjectRoleIDs {
				subjectIDs = append(subjectIDs, id)
			}
			_, err = o.QueryTable(new(models.RBACUserRoleAssignment)).
				Filter("role_id__in", subjectIDs).
				All(&assigned)
			if err != nil {
				return nil, err
			}
			currentUserIDs := map[string]bool{}
			for _, a := range assigned {
				if a.ValidFrom != nil && a.ValidFrom.After(now) {
					continue
				}
				if a.ValidUntil != nil && a.ValidUntil.Before(now) {
					continue
				}
				currentUserIDs[a.UserID] = true
			}

			userAlreadyCounted := currentUserIDs[userID]
			wouldReceiveSubject := false
			for id := range proposedRoleIDs {
				if subjectRoleIDs[id] {
					wouldReceiveSubject = true
					break
				}
			}
			projectedCount := len(currentUserIDs)
			if wouldReceiveSubject && !userAlreadyCounted {
				projectedCount++
			}

			if projectedCount > int(maxUsers) {
				names := joinRoleNames(subjectIDs, roleNameMap)
				violations = append(violations, "角色 "+names+" 最多只能分配给 "+strconv.Itoa(int(maxUsers))+" 个用户")
			}
		}
	}

	return violations, nil
}

func joinRoleNames(roleIDs []string, roleNameMap map[string]string) string {
	names := make([]string, 0, len(roleIDs))
	for _, id := range roleIDs {
		name := roleNameMap[id]
		if name == "" {
			name = id
		}
		names = append(names, name)
	}
	return strings.Join(names, "、")
}
